crates/services/consensus_module/poa/src/sync.rs

```diff
@@ -0,0 +1,563 @@
+use std::time::Duration;
+
+use fuel_core_services::{
+    stream::BoxStream,
+    RunnableService,
+    RunnableTask,
+    StateWatcher,
+};
+use fuel_core_types::fuel_types::BlockHeight;
+use tokio::sync::watch;
+use tokio_stream::StreamExt;
+
+use crate::deadline_clock::{
+    DeadlineClock,
+    OnConflict,
+};
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum SyncState {
+    NotSynced,
+    Synced,
+}
+
+impl SyncState {
+    pub fn from_config(
+        min_connected_reserved_peers: usize,
+        time_until_synced: Duration,
+    ) -> SyncState {
+        if min_connected_reserved_peers == 0 && time_until_synced == Duration::ZERO {
+            SyncState::Synced
+        } else {
+            SyncState::NotSynced
+        }
+    }
+}
+
+pub struct SyncTask {
+    min_connected_reserved_peers: usize,
+    time_until_synced: Duration,
+    peer_connections_stream: BoxStream<usize>,
+    block_stream: BoxStream<BlockHeight>,
+    state_sender: watch::Sender<SyncState>,
+    inner_state: InnerSyncState,
+    timer: DeadlineClock,
+}
+
+impl SyncTask {
+    pub fn new(
+        peer_connections_stream: BoxStream<usize>,
+        min_connected_reserved_peers: usize,
+        time_until_synced: Duration,
+        block_stream: BoxStream<BlockHeight>,
+        state_sender: watch::Sender<SyncState>,
+        block_height: BlockHeight,
+    ) -> Self {
+        let inner_state = InnerSyncState::from_config(
+            min_connected_reserved_peers,
+            time_until_synced,
+            block_height,
+        );
+        let timer = DeadlineClock::new();
+
+        Self {
+            peer_connections_stream,
+            min_connected_reserved_peers,
+            time_until_synced,
+            block_stream,
+            state_sender,
+            inner_state,
+            timer,
+        }
+    }
+
+    pub async fn run(&mut self, watcher: &mut StateWatcher) -> anyhow::Result<bool> {
```

Comment:
I see that you want to test each state transition separately in unit tests, but it seems it would be much simpler to have all transition code inside of the `SyncTask`. Maybe even all logic inside the `run` function(or you can split it into internal functions).

In the unit test, you can create a `SyncTask` instance and test each iteration of the `run` funciton.

Reply:
I have moved logic from `run()` method to `RunnableTask::run()` method, since having both didn't make much sense. Apart from that the code is already split into smaller encapsulated methods.